File: src/main/java/raylras/zen/langserver/provider/HoverProvider.java
package raylras.zen.langserver.provider;

import org.antlr.v4.runtime.tree.ParseTree;
import org.eclipse.lsp4j.Hover;
import org.eclipse.lsp4j.HoverParams;
import org.eclipse.lsp4j.MarkupContent;
import org.eclipse.lsp4j.MarkupKind;
import raylras.zen.bracket.BracketHandlerEntry;
import raylras.zen.bracket.BracketHandlerService;
import raylras.zen.code.Visitor;
import raylras.zen.code.parser.ZenScriptParser.BracketHandlerExprContext;
import raylras.zen.langserver.Document;
import raylras.zen.util.CSTNodes;
import raylras.zen.util.Position;
import raylras.zen.util.Ranges;

import java.util.Deque;
import java.util.concurrent.CompletableFuture;

public class HoverProvider {

    private HoverProvider() {}

    public static CompletableFuture<Hover> hover(Document doc, HoverParams params) {
        return doc.getUnit().map(unit -> CompletableFuture.supplyAsync(() -> {
            Position cursor = Position.of(params.getPosition());
            Deque<ParseTree> cstStack = CSTNodes.getCstStackAtPosition(unit.getParseTree(), cursor);
            HoverVisitor visitor = new HoverVisitor(unit.getEnv().getBracketHandlerService());
            for (ParseTree cst : cstStack) {
                Hover hover = cst.accept(visitor);
                if (hover != null) {
                    return hover;
                }
            }
            return null;
        })).orElseGet(HoverProvider::empty);
    }

    public static CompletableFuture<Hover> empty() {
        return CompletableFuture.completedFuture(null);
    }

    private static final class HoverVisitor extends Visitor<Hover> {

        private final BracketHandlerService brackets;

        private HoverVisitor(BracketHandlerService brackets) {
            this.brackets = brackets;
        }

        @Override
        public Hover visitBracketHandlerExpr(BracketHandlerExprContext ctx) {
            BracketHandlerEntry entry = brackets.queryEntryDynamic(ctx.raw().getText());
            StringBuilder builder = new StringBuilder();
            entry.ifPresent("_name", name -> {
                builder.append("#### ");
                builder.append(name);
                builder.append("\n\n");
            });
            entry.ifPresent("_icon", icon -> {
                String img = "![img](data:image/png;base64," + icon + ")";
                builder.append(img);
                builder.append("\n\n");
            });
            Hover hover = toHover(builder.toString());
            hover.setRange(Ranges.toLspRange(ctx));
            return hover;
        }

        private static Hover toHover(String text) {
            return new Hover(new MarkupContent(MarkupKind.MARKDOWN, text));
        }

        private static String toCodeBlock(String text) {
            return String.format("```zenscript\n%s\n```\n", text);
        }
    }

}

File: language-server/src/main/kotlin/raylras/intellizen/languageserver/provider/HoverProvider.kt
package raylras.intellizen.languageserver.provider

import org.antlr.v4.runtime.tree.RuleNode
import org.eclipse.lsp4j.Hover
import org.eclipse.lsp4j.HoverParams
import org.eclipse.lsp4j.MarkupContent
import org.eclipse.lsp4j.MarkupKind
import raylras.intellizen.CompilationUnit
import raylras.intellizen.Visitor
import raylras.intellizen.brackets.BracketHandlers
import raylras.intellizen.languageserver.util.toLspRange
import raylras.intellizen.languageserver.util.toTextPosition
import raylras.intellizen.parser.ZenScriptParser.BracketHandlerExprContext
import raylras.intellizen.util.getCstStackAt
import raylras.intellizen.util.textRange
import java.awt.image.BufferedImage
import java.io.ByteArrayOutputStream
import javax.imageio.ImageIO
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

object HoverProvider {
    fun hover(unit: CompilationUnit, params: HoverParams): Hover? {
        val cursor = params.position.toTextPosition()
        val cstStack = unit.parseTree.getCstStackAt(cursor)
        val visitor = HoverVisitor(unit)
        for (cst in cstStack) {
            cst.accept(visitor)?.let {
                return it
            }
        }
        return null
    }
}

private class HoverVisitor(private val unit: CompilationUnit) : Visitor<Hover?>() {
    override fun visitBracketHandlerExpr(ctx: BracketHandlerExprContext): Hover? {
        val expr = ctx.raw().text
        return BracketHandlers.getIconRemote(expr).fold(
            onSuccess = {
                it?.let { "![img](data:image/png;base64,${resize(it, 128)})" }
            },
            onFailure = {
                it.message
            }
        )?.toHover()?.apply {
            range = ctx.textRange.toLspRange()
        }
    }

    override fun visitChildren(node: RuleNode): Hover? {
        return null
    }
}

private fun String.toHover(): Hover = Hover(MarkupContent(MarkupKind.MARKDOWN, this))

private fun String.toCodeBlock() = """
    ```zenscript
    $this
    ```
""".trimIndent()

@OptIn(ExperimentalEncodingApi::class)
private fun resize(image64: String, size: Int): String {
    val input: BufferedImage = Base64.decode(image64).let {
        ImageIO.read(it.inputStream())
    }
    val output: BufferedImage = BufferedImage(size, size, input.type).apply {
        createGraphics().drawImage(input, 0, 0, size, size, null)
    }
    val outputStream = ByteArrayOutputStream().apply {
        ImageIO.write(output, "png", this)
    }
    return Base64.encode(outputStream.toByteArray())
}
